### ISSUE TITLE
`cabal-install` fails after upgrade to new `process`

### ISSUE THREAD

**Issue by hololeap** [gentoo-haskell/gentoo-haskell#1377]
With `process` now being upgradable (see 6212e01ac29233ea686d0655cd416adfdae49cb2), there is a snag when a package (such as `cabal-install`) pulls in something that built with `process-1.6.16.0` and _also_ pulls in `cabal` which comes bundled using `process-1.6.13.2`. For those using `ghc-8.10` and `ghc-9.0`, they can upgrade to/rebuild `cabal-3.6.3.0` and then run `haskell-updater`. For those using `ghc-9.2`, the current setup does not allow rebuilding that version of `cabal`.

Several solutions are being considered, but currently the least painful option seems to be:

- In a revision bump, give the packages that need to be rebuilt a slightly different version in their `.cabal` then those installed through ghc. For instance, if `haskeline-0.8.2` is installed through ghc, we can make the version installed via portage `haskeline-0.8.2.0.1`

Other possible solutions include:

- Patch `dev-lang/ghc` to use the new version of `process`
  _Cons:_ A big hassle and would force everyone to rebuild ghc
- Re-work the ghc bootstrap process so that core haskell packages are built afterwards using portage
  _Cons:_ Most likely difficult, if it's even possible at all

<details>
<summary>Error log
</summary>

```
[131 of 178] Compiling Distribution.Client.Manpage ( src/Distribution/Client/Manpage.hs, dist/build/cabal/cabal-tmp/Distribution/Client/Manpage.dyn_o )

src/Distribution/Client/Manpage.hs:57:44: error:
    • Couldn't match expected type ‘process-1.6.13.2:System.Process.Common.CreateProcess’
                  with actual type ‘Process.CreateProcess’
      NB: ‘process-1.6.13.2:System.Process.Common.CreateProcess’
            is defined in ‘System.Process.Common’ in package ‘process-1.6.13.2’
          ‘Process.CreateProcess’
            is defined in ‘System.Process.Common’ in package ‘process-1.6.16.0’
    • In the first argument of ‘createProcess’, namely
        ‘(Process.proc cmd args)
           {Process.std_in = Process.CreatePipe,
            Process.std_out = Process.Inherit,
            Process.std_err = Process.Inherit}’
      In a stmt of a 'do' block:
        (mb_in, _, _, ph) <- createProcess
                               (Process.proc cmd args)
                                 {Process.std_in = Process.CreatePipe,
                                  Process.std_out = Process.Inherit,
                                  Process.std_err = Process.Inherit}
      In the expression:
        do let cmd = "man"
               args = ...
           (mb_in, _, _, ph) <- createProcess
                                  (Process.proc cmd args)
                                    {Process.std_in = Process.CreatePipe,
                                     Process.std_out = Process.Inherit,
                                     Process.std_err = Process.Inherit}
           for_ mb_in $ \ hin -> do ...
           ec <- Process.waitForProcess ph
           ....
   |
57 |         (mb_in, _, _, ph) <- createProcess (Process.proc cmd args)
   |                                            ^^^^^^^^^^^^^^^^^^^^^^^...

src/Distribution/Client/Manpage.hs:69:38: error:
    • Couldn't match expected type ‘Process.ProcessHandle’
                  with actual type ‘process-1.6.13.2:System.Process.Common.ProcessHandle’
      NB: ‘Process.ProcessHandle’
            is defined in ‘System.Process.Common’ in package ‘process-1.6.16.0’
          ‘process-1.6.13.2:System.Process.Common.ProcessHandle’
            is defined in ‘System.Process.Common’ in package ‘process-1.6.13.2’
    • In the first argument of ‘Process.waitForProcess’, namely ‘ph’
      In a stmt of a 'do' block: ec <- Process.waitForProcess ph
      In the expression:
        do let cmd = "man"
               args = ...
           (mb_in, _, _, ph) <- createProcess
                                  (Process.proc cmd args)
                                    {Process.std_in = Process.CreatePipe,
                                     Process.std_out = Process.Inherit,
                                     Process.std_err = Process.Inherit}
           for_ mb_in $ \ hin -> do ...
           ec <- Process.waitForProcess ph
           ....
   |
69 |         ec <- Process.waitForProcess ph
   |                                      ^^
```

</details>

**Comment by hololeap** [comment]
It looks like `haskeline` is affected as well and will need to be rebuilt by everyone. Once I test it on my end, I will push revbumps that do not include `CABAL_CORE_LIB_GHC_PV` for those two packages so that everyone rebuilds. This will also be a good test to see if anything breaks when `CABAL_CORE_LIB_GHC_PV` is removed. We might want to get rid of it completely if all it does is prevent people from rebuilding.

**Comment by hololeap** [comment]
See: 6212e01ac29233ea686d0655cd416adfdae49cb2

**Comment by hololeap** [comment]
Here's the eclass docs for `CABAL_CORE_LIB_GHC_PV`:

https://github.com/gentoo-haskell/gentoo-haskell/blob/a0a8595362c8e9e1b8ea729d74f8533d5eff15d8/eclass/haskell-cabal.eclass#L612-L633

https://github.com/gentoo-haskell/gentoo-haskell/blob/a0a8595362c8e9e1b8ea729d74f8533d5eff15d8/eclass/ghc-package.eclass#L229-L250

**Comment by hololeap** [comment]
This presents with an interesting problem. _If_ we upgrade a core package for a bug fix and _if_ other core packages depend on it, we need a way to rebuild them. The problem lies in package collisions with the core packages installed in `ghc`. I see a couple ways around this:

1. Don't build the core packages when building `ghc`, but build them afterwards through portage.
2. Give the packages a slightly different version (through `.cabal`) then those installed through `ghc`. For instance, if `haskeline-0.8.2` is installed through `ghc`, we can make the version installed via portage `haskeline-0.8.2.0.1`

Option 2 may serve as workaround until option 1 gets worked out (if it's even possible).

**Comment by amano-kenji** [comment]
Obviously, option 1 is the permanent solution. With option 1, you remove all the future issues.

**Comment by hololeap** [comment]
@amano-kenji getting those bugs fixed is going to be very difficult as it requires a new EAPI from the gentoo council and an implementation in portage. The best we have right now is the use of an external tool (`haskell-updater`), because there are fundamental differences between how GHC's ABI is engineered and what portage is engineered to handle.

Solving option 1 wouldn't allow us to get rid of `haskell-updater`. The only thing that will allow that would be some mechanism within portage that would somehow allow portage to "understand" GHC's ABI and know how to fix breakages. This isn't a high priority for the main Gentoo devs, so I wouldn't recommend holding your breath waiting for it.

That said, we _can_ work on making `gentoo-haskell` a less painful experience even without direct support from portage. If you're running into issues rebuilding, feel free to ask questions here or on IRC (see the `README.md` for details on our IRC channel if you haven't already.)

---

- It's a hack, but I generally get through updates with the following:
  ```bash
  until emerge -DUu @world; do
      haskell-updater -- --ignore-built-slot-operator-deps=y || break
  done
  ```
- You can also sync `::haskell` less often so you don't have to deal with haskell breakages as often.
- Feel free to ask questions and report bugs
- Using `haskell-updater` _is_ annoying but there isn't a better option at the moment. Familiarize yourself with GHC's ABI to see what makes it unique and why portage has a hard time with it. Any ideas are appreciated but please understand it is a difficult problem.

**Comment by amano-kenji** [comment]
I have used `emerge -uDN @world`. What is the difference between `-uDN` and `-uUD`?

**Comment by hololeap** [comment]
@amano-kenji can you open this as a new issue and include logs of the error you're running into?

From `man emerge`:

```
--changed-use, -U
        Tells  emerge  to include installed packages where USE flags have changed since installation. This option also implies the --selective option. Unlike --newuse, the --changed-use option does not trigger reinstallation
        when flags that the user has not enabled are added or removed.
```

```
--newuse, -N
        Tells emerge to include installed packages where USE flags have changed since compilation. This option also implies the --selective option.  USE flag changes include:

        A USE flag was added to a package.  A USE flag was removed from a package.  A USE flag was turned on for a package.  A USE flag was turned off for a package.

        USE flags may be toggled by your profile as well as your USE and package.use settings. If you would like to skip rebuilds for which  disabled  flags  have  been  added  to  or  removed  from  IUSE,  see  the  related
        --changed-use option. If you would like to skip rebuilds for specific packages, see the --exclude option.
```

The key part is 

> Unlike --newuse, the --changed-use option does not trigger reinstallation when flags that the user has not enabled are added or removed.

**Comment by amano-kenji** [comment]
I just synchronized my local copy of haskell overlay, and this issue still exists. Are you sure a new issue is needed?

**Comment by hololeap** [comment]
If it's failing to build `dev-haskell/cabal-install` then you should just need to rebuild `dev-haskell/cabal` and then run `haskell-updater`.

If you're running into other errors, then you should open it up as a new issue.

**Comment by amano-kenji** [comment]
> If it's failing to build dev-haskell/cabal-install then you should just need to rebuild dev-haskell/cabal and then run haskell-updater.

How the hell did you figure this out? I couldn't figure this out from errors alone. I am frsutrated by the fact that emerge can't figure out that it needed to rebuild cabal before building cabal-install.

Perhaps, haskell-updater could have built cabal automatically for me?

If tools don't take care of build orders for me, I'm lost. If haskell-updater doesn't catch errors, I probably can't. Haskell language is beautiful, but haskell tooling has been awkward.

**Comment by hololeap** [comment]
@amano-kenji See: https://github.com/gentoo-haskell/gentoo-haskell/blob/master/projects/doc/maintaining-tips.rst#on-problems-with-slotting-haskell-libraries

By making `process` upgradable, I introduced a de-facto slot, because now there can be two versions of `process` installed on the system; the one that comes bundled with ghc and one installed via portage.

> With `process` now being upgradable (see [6212e01](https://github.com/gentoo-haskell/gentoo-haskell/commit/6212e01ac29233ea686d0655cd416adfdae49cb2)), there is a snag when a package (such as `cabal-install`) pulls in something that built with `process-1.6.16.0` and with `cabal` which comes bundled using `process-1.6.13.2`.

There was some unexpected fallout from the decision to make `process` upgradable, but I think it's more or less been managed at this point.

**Comment by amano-kenji** [comment]
I would really like ghc-9.2.1 and above for NoFieldSelector and other record-related extensions.

Since cabal's revision has been bumped up, I think the issue can be closed now?

**Comment by hololeap** [comment]
This should be fixed with fb9738bbe1df2cf0900c1c5693c5ac6f7aa0db08.
